packages/dso-toolkit/gulp/build-svg-spritesheet.js

```diff
@@ -13,135 +13,131 @@ import { sassTransformer } from "./transformers/sass.transformer.js";
+            const ast = parse(style); // Parse the CSS into an abstract syntax tree (AST)
+
+            // Extract color variants for each CSS rule
+            const variants = ast.stylesheet.rules
+              .filter((rule) => rule.type === "rule") // Keep only CSS rules (ignore comments, etc.)
+              .flatMap((rule) => {
+                const colorDeclaration = rule.declarations.find((d) => d.property === "color");
+                if (!colorDeclaration) return [];
+
+                const color = colorDeclaration.value;
+                return rule.selectors
+                  .map((selector) => {
+                    if (selector === id) {
+                      return { selector: "default", color }; // Mark as default variant
                     }
 
-                    if (
-                      s.startsWith(`${id}:`) &&
-                      !v.some((variantColor) => variantColor.selector === s.substr(`${id}:`.length))
-                    ) {
-                      return current.concat({
-                        color,
-                        selector: s.substr(`${id}:`.length),
-                      });
+                    if (selector.startsWith(`${id}:`)) {
+                      return { selector: selector.slice(id.length + 1), color }; // Extract the variant name
                     }
 
-                    return current;
-                  }, []);
-
-                  v.push(...variantColors);
+                    return null;
+                  })
+                  .filter(Boolean); // Remove null entries
+              });
 
-                  return v;
-                }, []);
+            return { id, variants };
+          });
 
-              return { id, variants };
-            }),
-          );
+          resolve(parsedStyles);
         }),
       );
   });
+}
 
-  return new Promise((resolve, reject) => {
-    gulp
-      .src(`${iconsPath}/*.svg`)
-      .pipe(prettier({ plugins: ["@prettier/plugin-xml"] }))
-      .pipe(
-        svgstore({
-          inlineSvg: true,
-        }),
-      )
-      .pipe(
-        cheerio({
-          run($) {
-            const canvas = 32;
-            const gutter = 10;
-            const symbols = $("symbol").toArray();
-            const positions = symbols.reduce((position, element) => {
-              const symbol = $(element);
-              const id = symbol.attr("id");
-
-              const stylesheet = stylesheets.find((s) => s.id === id);
+async function buildSprite(iconsWithStyles = false) {
+  const stylesheets = await parseStylesheets(); // Parse stylesheets within the function
+
+  return gulp
+    .src(`${iconsPath}/*.svg`)
+    .pipe(prettier({ plugins: ["@prettier/plugin-xml"] }))
+    .pipe(
+      svgstore({
+        inlineSvg: true,
+      }),
+    )
+    .pipe(
+      cheerio({
+        run($) {
+          const symbols = $("symbol").toArray();
+          let position = 0;
+
+          symbols.forEach((symbol) => {
```

Comment:
Kan je wat vertellen over de symbols.reduce() met het aantal posities als return type, versus een symbols.forEach() met de addition assigment operator? Waarom heeft die laatste jouw voorkeur?

Overigens is `Array.prototype.forEach()` een suboptimale iteratie routine. Hij is traag, kan niet worden gestopt (`break`) of overgeslagen (`continue`).

Persoonlijk mijd ik 'm altijd. Ik kan geen reden bedenken om `Array.prototype.forEach()` te gebruiken. Itereren over lijstjes is primaire functionaliteit van een programmeertaal, en gelukkig heeft JavaScript die geimplementeerd met `for ... of` :)

Zie ook https://jsperf.app/for-for-of-for-in-foreach-comparison en https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...of

Reply:
### Waarom forEach() hier de voorkeur heeft
**Leesbaarheid in een complexe workflow:**

Een forEach() is makkelijker te begrijpen in een situatie waarin het resultaat niet alleen het ophogen van een waarde is, maar ook waar side-effects optreden zoals het manipuleren van DOM-elementen of het aanroepen van complexe functies (view en group in de code).
Het toevoegen van een position-wijziging voelt het voor mij meer intuïtief aan wanneer deze direct volgt na een operatie die afhankelijk is van de huidige position.

**Intentie is mutatie, niet aggregatie:**

ForEach() is meer geschikt als de intentie het uitvoeren van operaties op items in een lijst is, terwijl reduce() beter past bij puur aggregatiewerk (zoals het berekenen van een totaal). In dit geval doe je beide maar met de nadruk op het manipuleren van items.

Dat gezegd hebbende ben ik ook voorstander van een `for of` loop. Bij het schrijven van de code koos eerst voor het gebruik van forEach omwille van de programmeer 'snelheid' om het issue af te ronden, echter zie ik nu in dat ik deze had moeten refactoren zodat de code sneller de svg's kan renderen.

Reply:
Check, in dat geval denken we er hetzelfde over :)

Overigens was mijn overweging niet eens per se uit performance. Ik ben gewoon geen fan van `Array.prototype.forEach()` 😄